File: unii-app/lib/data/models/post.dart
class MediaAsset {
  const MediaAsset({
    required this.id,
    required this.mediaType,
    required this.url,
    this.thumbnailUrl,
    this.duration,
    this.sortOrder = 0,
  });

  factory MediaAsset.fromJson(Map<String, dynamic> json) => MediaAsset(
    id: (json['id'] as num).toInt(),
    mediaType: json['media_type'] as String,
    url: json['url'] as String,
    thumbnailUrl: json['thumbnail_url'] as String?,
    duration: (json['duration'] as num?)?.toInt(),
    sortOrder: (json['sort_order'] as num?)?.toInt() ?? 0,
  );

  final int id;
  final String mediaType;
  final String url;
  final String? thumbnailUrl;
  final int? duration;
  final int sortOrder;

  bool get isImage => mediaType == 'image';
}

class Post {
  const Post({
    required this.id,
    required this.authorId,
    required this.authorUsername,
    required this.visibility,
    required this.likeCount,
    required this.commentCount,
    required this.likedByMe,
    required this.media,
    this.authorNickname,
    this.authorAvatarUrl,
    this.title,
    this.content,
    this.teamId,
    this.activityId,
    this.postType = 0,
    this.createdAt,
  });

  factory Post.fromJson(Map<String, dynamic> json) => Post(
    id: (json['id'] as num).toInt(),
    authorId: (json['author_id'] as num).toInt(),
    authorUsername: json['author_username'] as String,
    authorNickname: json['author_nickname'] as String?,
    authorAvatarUrl: json['author_avatar_url'] as String?,
    visibility: json['visibility'] as String,
    likeCount: (json['like_count'] as num).toInt(),
    commentCount: (json['comment_count'] as num).toInt(),
    likedByMe: (json['liked_by_me'] as bool?) ?? false,
    media: ((json['media'] as List<dynamic>?) ?? const [])
        .map((e) => MediaAsset.fromJson(e as Map<String, dynamic>))
        .toList(growable: false),
    title: json['title'] as String?,
    content: json['content'] as String?,
    teamId: (json['team_id'] as num?)?.toInt(),
    activityId: (json['activity_id'] as num?)?.toInt(),
    postType: (json['post_type'] as num?)?.toInt() ?? 0,
    createdAt: json['created_at'] as String?,
  );

  final int id;
  final int authorId;
  final String authorUsername;
  final String? authorNickname;
  final String? authorAvatarUrl;
  final int? teamId;
  final int? activityId;
  final int postType;
  final String? title;
  final String? content;
  final String visibility;
  final int likeCount;
  final int commentCount;
  final bool likedByMe;
  final List<MediaAsset> media;
  final String? createdAt;

  String get authorDisplayName =>
      (authorNickname?.isNotEmpty ?? false) ? authorNickname! : authorUsername;

  Post copyWith({bool? likedByMe, int? likeCount, int? commentCount}) => Post(
    id: id,
    authorId: authorId,
    authorUsername: authorUsername,
    authorNickname: authorNickname,
    authorAvatarUrl: authorAvatarUrl,
    teamId: teamId,
    activityId: activityId,
    postType: postType,
    title: title,
    content: content,
    visibility: visibility,
    likeCount: likeCount ?? this.likeCount,
    commentCount: commentCount ?? this.commentCount,
    likedByMe: likedByMe ?? this.likedByMe,
    media: media,
    createdAt: createdAt,
  );
}

class PostComment {
  const PostComment({
    required this.id,
    required this.postId,
    required this.userId,
    required this.username,
    required this.content,
    this.nickname,
    this.avatarUrl,
    this.parentId,
    this.createdAt,
  });

  factory PostComment.fromJson(Map<String, dynamic> json) => PostComment(
    id: (json['id'] as num).toInt(),
    postId: (json['post_id'] as num).toInt(),
    userId: (json['user_id'] as num).toInt(),
    username: json['username'] as String,
    content: json['content'] as String,
    nickname: json['nickname'] as String?,
    avatarUrl: json['avatar_url'] as String?,
    parentId: (json['parent_id'] as num?)?.toInt(),
    createdAt: json['created_at'] as String?,
  );

  final int id;
  final int postId;
  final int userId;
  final String username;
  final String? nickname;
  final String? avatarUrl;
  final int? parentId;
  final String content;
  final String? createdAt;

  String get displayName =>
      (nickname?.isNotEmpty ?? false) ? nickname! : username;
}

class LikeToggleResult {
  const LikeToggleResult({required this.liked, required this.likeCount});

  factory LikeToggleResult.fromJson(Map<String, dynamic> json) =>
      LikeToggleResult(
        liked: json['liked'] as bool,
        likeCount: (json['like_count'] as num).toInt(),
      );

  final bool liked;
  final int likeCount;
}

File: unii-server/src/dto/post.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

use crate::model::post::{MediaRow, PostRow};

#[derive(Debug, Deserialize)]
pub struct CreatePostReq {
    pub title: Option<String>,
    pub content: Option<String>,
    /// `public` | `private` (W5 only `public` is meaningful in feed; private posts
    /// are visible to author only).
    pub visibility: String,
    pub team_id: Option<i64>,
    /// IDs of `media_files` rows previously uploaded via `/media/upload`.
    /// Server attaches them by setting `owner_type='post'` + `owner_id=<post.id>`.
    #[serde(default)]
    pub media_ids: Vec<i64>,
}

#[derive(Debug, Deserialize)]
pub struct CreateCommentReq {
    pub content: String,
    pub parent_id: Option<i64>,
}

#[derive(Debug, Deserialize)]
pub struct FeedQuery {
    pub before_id: Option<i64>,
    pub limit: Option<i32>,
}

#[derive(Debug, Serialize)]
pub struct PostView {
    pub id: i64,
    pub author_id: i64,
    pub author_username: String,
    pub author_nickname: Option<String>,
    pub author_avatar_url: Option<String>,
    pub team_id: Option<i64>,
    pub activity_id: Option<i64>,
    pub post_type: i16,
    pub title: Option<String>,
    pub content: Option<String>,
    pub visibility: String,
    pub like_count: i32,
    pub comment_count: i32,
    pub created_at: Option<DateTime<Utc>>,
    pub media: Vec<MediaView>,
    pub liked_by_me: bool,
}

#[derive(Debug, Serialize)]
pub struct MediaView {
    pub id: i64,
    pub media_type: String,
    pub url: String,
    pub thumbnail_url: Option<String>,
    pub duration: Option<i32>,
    pub sort_order: i16,
}

impl From<&MediaRow> for MediaView {
    fn from(r: &MediaRow) -> Self {
        Self {
            id: r.id,
            media_type: r.media_type.clone(),
            url: r.url.clone(),
            thumbnail_url: r.thumbnail_url.clone(),
            duration: r.duration,
            sort_order: r.sort_order.unwrap_or(0),
        }
    }
}

#[derive(Debug, Serialize)]
pub struct CommentView {
    pub id: i64,
    pub post_id: i64,
    pub user_id: i64,
    pub username: String,
    pub nickname: Option<String>,
    pub avatar_url: Option<String>,
    pub parent_id: Option<i64>,
    pub content: String,
    pub created_at: Option<DateTime<Utc>>,
}

#[derive(Debug, Serialize)]
pub struct LikeToggleResp {
    pub liked: bool,
    pub like_count: i32,
}

/// Joined row for repo-level comment listing.
#[derive(Debug, Clone, sqlx::FromRow)]
pub struct CommentJoinRow {
    pub id: i64,
    pub post_id: i64,
    pub user_id: i64,
    pub username: String,
    pub nickname: Option<String>,
    pub avatar_url: Option<String>,
    pub parent_id: Option<i64>,
    pub content: String,
    pub created_at: Option<DateTime<Utc>>,
}

impl From<CommentJoinRow> for CommentView {
    fn from(r: CommentJoinRow) -> Self {
        Self {
            id: r.id,
            post_id: r.post_id,
            user_id: r.user_id,
            username: r.username,
            nickname: r.nickname,
            avatar_url: r.avatar_url,
            parent_id: r.parent_id,
            content: r.content,
            created_at: r.created_at,
        }
    }
}

/// Joined row for repo-level feed listing.
#[derive(Debug, Clone, sqlx::FromRow)]
pub struct PostJoinRow {
    pub id: i64,
    pub author_id: i64,
    pub author_username: String,
    pub author_nickname: Option<String>,
    pub author_avatar_url: Option<String>,
    pub team_id: Option<i64>,
    pub activity_id: Option<i64>,
    pub post_type: i16,
    pub title: Option<String>,
    pub content: Option<String>,
    pub visibility: String,
    pub like_count: i32,
    pub comment_count: i32,
    pub created_at: Option<DateTime<Utc>>,
}

pub fn build_post_view(row: PostJoinRow, media: Vec<MediaView>, liked_by_me: bool) -> PostView {
    PostView {
        id: row.id,
        author_id: row.author_id,
        author_username: row.author_username,
        author_nickname: row.author_nickname,
        author_avatar_url: row.author_avatar_url,
        team_id: row.team_id,
        activity_id: row.activity_id,
        post_type: row.post_type,
        title: row.title,
        content: row.content,
        visibility: row.visibility,
        like_count: row.like_count,
        comment_count: row.comment_count,
        created_at: row.created_at,
        media,
        liked_by_me,
    }
}

pub fn build_post_view_from_row(
    row: &PostRow,
    author_username: String,
    author_nickname: Option<String>,
    author_avatar_url: Option<String>,
    media: Vec<MediaView>,
    liked_by_me: bool,
) -> PostView {
    PostView {
        id: row.id,
        author_id: row.author_id,
        author_username,
        author_nickname,
        author_avatar_url,
        team_id: row.team_id,
        activity_id: row.activity_id,
        post_type: row.post_type,
        title: row.title.clone(),
        content: row.content.clone(),
        visibility: row.visibility.clone(),
        like_count: row.like_count,
        comment_count: row.comment_count,
        created_at: row.created_at,
        media,
        liked_by_me,
    }
}
